### PR TITLE
Direct TCP | Address socket leak

### PR DESCRIPTION
**Pull request by David-Noble-at-work**
PR is ready for review with these TODOs before we merge.

- Get a pass or address test failures documented below. 
  See [analysis of current test failures](#current-test-failures).

- Complete customer ready JMH test runs on 4 and 8 proc containers.
  Comparative JMH test runs were completed here were run on a Standard F16s_v2 (16 vcpus, 32 GiB memory) VM.

#### Table of contents

- [Azure:master vs David-Noble-at-work:isssue/cosmos-4.X/direct-tcp/socket-leak](#master-vs-pr)
    - Socket utilization
        - [Concurrency: 16](#socket-utilization-concurrency-16)
        - [Concurrency: 32](#socket-utilization-concurrency-32)
        - [Concurrency: 64](#socket-utilization-concurrency-64)
        - [Concurrency: 128](#socket-utilization-concurrency-128)
        - [Concurrency: 256](#socket-utilization-concurrency-256)
        - [Concurrency: 512](#socket-utilization-concurrency-512)
        - [Concurrency: 1024](#socket-utilization-concurrency-1024)
        - [Concurrency: 2048](#socket-utilization-concurrency-2048)
    - [Performance](#performance)
        - [Throughput](#throughput)
        - [Latency](#latency)
- [Socket utilization in development environments with this change in place](#socket-utilization-in-development-environments)
  - [macOS](#macos)
  - [Windows](#windows)
- [Other issues addressed while creating this PR](#other-issues-addressed)
- [Analysis of current test failures](#current-test-failures)

#### Azure:master vs. David-Noble-at-work:isssue/cosmos-4.X/direct-tcp/socket-leak <a href="#user-content-master-vs-pr" id="master-vs-pr">#</a>

Anecdotal evidence had suggested that socket leaks were more likely to occur in development environments than production environments. The systematic study presented here shows that socket leaks will occur in a production environments like this one as well:

```json
{
    "client": "Linux appserver-lin-3 5.0.0-1036-azure #38-Ubuntu SMP Sun Mar 22 21:27:21 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux",
    "vmSize": "Standard F16s_v2 (16 vcpus, 32 GiB memory)",
    "javaVersion": "openjdk version 1.8.0_252 (Zulu)",
    "benchmark": "ReadLatency",
    "consistencyLevel": "Eventual",
    "concurrency": [16, 32, 64, 128, 256, 1024, 2048],
    "documentCount": [1, 100000],
    "operationCount": "50,000,000"
}
```

This PR corrects this issue. Here we show that at concurrencies between 16 and 2048 the value of `maxChannels` is respected and--consequently--there are no socket leaks. At concurrency levels at or above 256 we observe some performance degradation. Reviewers will see that performance differences could be the result of maxing out on channels. At a concurrency of 2048, for example, reviewers will note that channel utilization is at 100% with 10 channels allocated for each endpoint. The result is a difference in performance of around 5-6% as well as an increase in average latency. Warmup times might also have increased a little, though we did not specifically measure it. (Hypothesis: Additional connections are now guaranteed to be opened one at a time and this increases our warmup time.)

All of this indicates that performance tuning using `RntbdTransportClient.Options` may be a fruitful next step in `RntbdTransportClient` performance work. We might specifically run a set of experiments designed to measure the performance/latency curves for select values of:

* `maxChannels`
* `maxRequestsPerChannel`
* `threadCount`, a new option doubled from its fixed value to 2 * processor-count in this PR.

Such experiments would be a good segue into work on developing option profiles suited to specific workloads (e.g., read, write, query, or mixed at specific concurrency levels). Customers could then make tradeoffs between Cosmos client performance and Cosmos client resource consumption (CPU, Memory, and Sockets).

###### Azure:master
image

###### David-Noble-at-work:isssue/cosmos-4.X/direct-tcp/socket-leak
image

###### Azure:master
image

###### David-Noble-at-work:isssue/cosmos-4.X/direct-tcp/socket-leak
image

#####Socket utilization : Concurrency : 64 <a href="#user-content-socket-utilization-concurrency-64" id="socket-utilization-concurrency-64">#</a>

###### Azure:master
image

###### David-Noble-at-work:isssue/cosmos-4.X/direct-tcp/socket-leak
image

###### Azure:master
image

###### David-Noble-at-work:isssue/cosmos-4.X/direct-tcp/socket-leak
image

###### Azure:master
image

###### David-Noble-at-work:isssue/cosmos-4.X/direct-tcp/socket-leak
image

###### Azure:master
image

###### David-Noble-at-work:isssue/cosmos-4.X/direct-tcp/socket-leak
image

###### Azure:master
image

###### David-Noble-at-work:isssue/cosmos-4.X/direct-tcp/socket-leak
image

###### Azure:master
image

###### David-Noble-at-work:isssue/cosmos-4.X/direct-tcp/socket-leak
image

#### Performance <a href="#user-content-performance" id="performance">#</a>

We utilized two performance test programs in these runs:

- Java SDK benchmark tool for measuring asynchronous ReadLatency performance.
- Our internal JMH-based benchmark tool for measuring blocking ReadLatency performance.

###### Throughput <a href="#user-content-throughput" id="throughput">#</a>
image
image

###### Latency <a href="#user-content-throughput" id="latency">#</a>
image

image

##### ReadLatency : macOS development environment, documentCount = [1, 100,000] <a href="#user-content-macos" id="macos">#</a>
image

```json
{
    "client": "Danoble-MBP-6.thenobles.us, Darwin Kernel Version 19.4.0 x86_64 i386 MacBookPro16,1",
    "javaVersion": "openjdk version 1.8.0_252 (AdoptOpenJDK)",
    "benchmark": "ReadLatency",
    "concurrency": 2048,
    "documentCount": [
        1,
        100000
    ]
}
```

##### ReadLatency : Windows development environment, documentCount = [1, 100,000] <a href="#user-content-windows" id="windows">#</a>
 
image

```json
{
    "client": "Danoble-WS19-1, Microsoft Windows Server 2019 Datacenter Version 10.0.17763",
    "javaVersion": "openjdk version 11.0.3 2019-04-16 LTS (Zulu)",
    "benchmark": "ReadLatency",
    "concurrency": 2048,
    "documentCount": [
        1,
        100000
    ]
}
```

#### Other issues addressed while creating this PR <a href="#user-content-other-issues-addressed" id="other-issues-addressed">#</a>

- `RntbdRequestTimer` now never prevents program termination.
   I discovered that by default that Netty's `HashedWheelTimer` creates a user thread by default and that this user thread is not guaranteed to stop running when the `HashedWheelTimer` is stopped. `RntbdRequestTimer` now creates a daemon thread using a custom thread factory: `RntbdThreadFactory`. Our custom thread factory is now also used in `RntbdClientChannelPool` to allocate a static daemon thread for closing a channel pool.

- `RntbdReporter` now always produces a useful stack trace

- `RntbdTransportClient.Options` now includes:

    `connectionAcquisitionTimeout`\
     If a channel cannot be acquired within this time interval, it is failed. The default value is `Duration.ZERO` and specifies that there is no connection acquisition timeout.

    `threadCount`\
    The number of IO threads to allocate to the `RntbdTransportClient` event loop group. The default value is `2 * Runtime.getRuntime().availableProcessors()`. This is a two-fold increase over the previous value which was not configurable. The new default value (which is the default value used by Netty) was found to consistently produce a small (<5%) performance boost.

- Added benchmark configuration option: `-reportingDirectory`
  This option causes a CSV reporter to be created instead of a console reporter. I would like to rethink the option name on another PR. Reviewers: Is this good enough for now?

- Removed a few low-value debug-level logger messages.

#### Analysis of current test failures <a href="#user-content-current-test-failures" id="current-test-failures">#</a>

All but one build failure are for the result of failures in one or more of these tests:
- `ConsistencyTests2::validateSessionTokenAsync`
- `ConsistencyTests2::validateSessionTokenWithDocumentNotFound`
- `ConsistencyTests2::validateSessionTokenWithPreConditionFailure`

One build failed due to this test failure:
- `UniqueIndexTest.insertWithUniqueIndex`

CI Task Insights reports that these failures have occurred "[multiple] times across 15 pipeline runs in the last 14 days."

##### Single_Region_Strong_Tcp:Strong_Tcp_Direct, MultiMaster_Single_Region:Direct
```
2020-05-03T16:44:54.9731158Z [ERROR] Tests run: 79, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 870.012 s <<< FAILURE! - in TestSuite
2020-05-03T16:44:54.9764800Z [ERROR] ConsistencyTests2::validateSessionTokenAsync(com.azure.cosmos.implementation.ConsistencyTests2)  Time elapsed: 480.014 s  <<< FAILURE!
2020-05-03T16:44:54.9774100Z org.testng.internal.thread.ThreadTimeoutException: Method com.azure.cosmos.implementation.ConsistencyTests2.validateSessionTokenAsync() didn't finish within the time-out 480000
2020-05-03T16:44:54.9775987Z 	at app//com.azure.cosmos.implementation.ConsistencyTests2.lambda$validateSessionTokenAsync$0(ConsistencyTests2.java:235)
2020-05-03T16:44:54.9788238Z 	at app//com.azure.cosmos.implementation.ConsistencyTests2.validateSessionTokenAsync(ConsistencyTests2.java:272)
```

##### Single_Region_Session:Single Region Direct, Single_Region_Session_2:Multimaster Single Region Direct

```
2020-05-03T16:44:21.4096557Z [ERROR] Tests run: 79, Failures: 2, Errors: 0, Skipped: 6, Time elapsed: 803.098 s <<< FAILURE! - in TestSuite
2020-05-03T16:44:21.4131385Z [ERROR] ConsistencyTests2::validateSessionTokenAsync(com.azure.cosmos.implementation.ConsistencyTests2)  Time elapsed: 480.009 s  <<< FAILURE!
2020-05-03T16:44:21.4133834Z org.testng.internal.thread.ThreadTimeoutException: Method com.azure.cosmos.implementation.ConsistencyTests2.validateSessionTokenAsync() didn't finish within the time-out 480000
2020-05-03T16:44:21.4144820Z 	at app//com.azure.cosmos.implementation.ConsistencyTests2.lambda$validateSessionTokenAsync$0(ConsistencyTests2.java:235)
2020-05-03T16:44:21.4146499Z 	at app//com.azure.cosmos.implementation.ConsistencyTests2.validateSessionTokenAsync(ConsistencyTests2.java:272)
2020-05-03T16:44:21.4147349Z 
2020-05-03T16:44:21.4150290Z [ERROR] ConsistencyTests2::validateSessionTokenWithDocumentNotFound(com.azure.cosmos.implementation.ConsistencyTests2)  Time elapsed: 20.558 s  <<< FAILURE!
2020-05-03T16:44:21.4154098Z java.lang.AssertionError: 
2020-05-03T16:44:21.4154735Z 
2020-05-03T16:44:21.4155411Z Expecting:
2020-05-03T16:44:21.4156076Z  <false>
2020-05-03T16:44:21.4156764Z to be equal to:
2020-05-03T16:44:21.4157435Z  <true>
2020-05-03T16:44:21.4158094Z but was not.
2020-05-03T16:44:21.4167664Z 	at com.azure.cosmos.implementation.ConsistencyTests2.validateSessionTokenWithDocumentNotFound(ConsistencyTests2.java:114)
2020-05-03T16:44:21.4168735Z 
```
##### MultiMaster_Multi_Region:Direct
```
2020-05-03T18:06:39.2863867Z [ERROR] Tests run: 79, Failures: 2, Errors: 0, Skipped: 6, Time elapsed: 773.709 s <<< FAILURE! - in TestSuite
2020-05-03T18:06:39.2914217Z [ERROR] ConsistencyTests2::validateSessionTokenAsync(com.azure.cosmos.implementation.ConsistencyTests2)  Time elapsed: 480.014 s  <<< FAILURE!
2020-05-03T18:06:39.2916257Z org.testng.internal.thread.ThreadTimeoutException: Method com.azure.cosmos.implementation.ConsistencyTests2.validateSessionTokenAsync() didn't finish within the time-out 480000
2020-05-03T18:06:39.2924420Z 	at app//com.azure.cosmos.implementation.ConsistencyTests2.lambda$validateSessionTokenAsync$0(ConsistencyTests2.java:235)
2020-05-03T18:06:39.2931550Z 	at app//com.azure.cosmos.implementation.ConsistencyTests2.validateSessionTokenAsync(ConsistencyTests2.java:272)
2020-05-03T18:06:39.2932367Z 
2020-05-03T18:06:39.2933305Z [ERROR] ConsistencyTests2::validateSessionTokenWithPreConditionFailure(com.azure.cosmos.implementation.ConsistencyTests2)  Time elapsed: 30.395 s  <<< FAILURE!
2020-05-03T18:06:39.2938156Z java.lang.AssertionError: 
2020-05-03T18:06:39.2938709Z 
2020-05-03T18:06:39.2942431Z Expecting:
2020-05-03T18:06:39.2952499Z  <false>
2020-05-03T18:06:39.2953245Z to be equal to:
2020-05-03T18:06:39.2953913Z  <true>
2020-05-03T18:06:39.2954390Z but was not.
2020-05-03T18:06:39.2964162Z 	at com.azure.cosmos.implementation.ConsistencyTests2.validateSessionTokenWithPreConditionFailure(ConsistencyTests2.java:105)
20
```

##### Single_Region_Session_2 Single_Region_Long:Long Integration Tests
```
2020-05-03T17:44:30.4500562Z [ERROR] UniqueIndexTest::insertWithUniqueIndex(com.azure.cosmos.rx.UniqueIndexTest)  Time elapsed: 30.067 s  <<< FAILURE!
2020-05-03T17:44:30.4502230Z org.testng.internal.thread.ThreadTimeoutException: Method com.azure.cosmos.rx.UniqueIndexTest.insertWithUniqueIndex() didn't finish within the time-out 30000
2020-05-03T17:44:30.4503311Z 	at app//com.azure.cosmos.rx.UniqueIndexTest.insertWithUniqueIndex(UniqueIndexTest.java:116)
2020-05-03T17:44:30.4503993Z 
```

### CODE REVIEW

File: eng/versioning/version_client.txt [hunk 1]
```diff
@@ -17,8 +17,8 @@ com.azure:azure-core-serializer-json-jackson;1.0.0-beta.1;1.0.0-beta.1
 com.azure:azure-core-test;1.2.0;1.3.0-beta.1
 com.azure:azure-core-tracing-opentelemetry;1.0.0-beta.4;1.0.0-beta.5
 com.azure:azure-cosmos;4.0.1-beta.2;4.0.1-beta.3
-com.azure:azure-cosmos-examples;4.0.1-beta.1;4.0.1-beta.1
-com.azure:azure-cosmos-benchmark;4.0.1-beta.1;4.0.1-beta.1
+com.azure:azure-cosmos-examples;4.0.1-beta.1;4.0.1-beta.3
```
Comment (JimSuplizio) [inline eng/versioning/version_client.txt:20]: @kushagraThapar 
com.azure:azure-cosmos-examples and azure-cosmos-benchmark are in the process is being removed as libraries. Further, they're intentionally not part of the Artifacts cosmos' [ci.yml](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/cosmos/ci.yml#L52) which makes them incapable of being released (this was intentional as these are never supposed to be released). There is no version change necessary here and it should be reverted.

Comment (kushagraThapar) [reply eng/versioning/version_client.txt:20]: @JimSuplizio - yes this is unintentional, should be reverted. 
@David-Noble-at-work - can we please revert the changes related to pom.xml ?

Comment (David-Noble-at-work) [reply eng/versioning/version_client.txt:20]: @kushagraThapar @JimSuplizio yep

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java [hunk 2]
```diff
@@ -667,12 +1329,24 @@ public void serialize(final RntbdClientChannelPool value, final JsonGenerator ge
...
             generator.writeNumberField("channelsAvailable", value.channelsAvailable());
             generator.writeNumberField("requestQueueLength", value.requestQueueLength());
-            generator.writeNumberField("usedDirectMemory", value.usedDirectMemory());
-            generator.writeNumberField("usedHeapMemory", value.usedHeapMemory());
             generator.writeEndObject();
             generator.writeEndObject();
         }
     }
 
+    private static class StacklessIllegalStateException extends IllegalStateException {
```
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java:1337]: thank you.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java [hunk 3]
```diff
@@ -411,48 +729,389 @@ private void close0() {
...
+//        final int channelsAcquired = this.channelsAcquired();
+//        final int channelCount = this.channels();
+//
+//        reportIssueUnless(logger, 0 <= channelCount && channelCount <= this.maxChannels, this,
+//            "expected channelCount in range [0, {}], not {}",
+//            this.maxChannels,
+//            channelCount,
+//            channelsAcquired);
+//
+//        return channelsAvailable;
```
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java:786]: I will remove this in my next PR which will wire up the `DirectConnectionConfig`. The change is made on my local branch right now.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java [hunk 4]
```diff
@@ -281,126 +439,286 @@ public String toString() {
...
-        final Channel first = super.pollChannel();
-
-        if (first == null) {
-            return null;
+    private void addTaskToPendingAcquisitionQueue(Promise<Channel> promise) {
+
+        this.ensureValidRunState();
+
+        if (logger.isDebugEnabled()) {
+            logger.debug("{}, {}, {}, {}, {}, {}",
```
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java:567]: Here's what I see in the code:
```
            logger.debug("{}, {}, {}, {}, {}, {}",
                OffsetDateTime.now(),
                this.remoteAddress(),
                this.channels(),
                this.channelsAcquired(),
                this.channelsAvailable(),
                this.requestQueueLength());
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java [hunk 5]
```diff
@@ -233,45 +349,87 @@ public boolean isClosed() {
...
             if (future.isSuccess()) {
-
-                this.decrementAndRunTaskQueue();
+                this.runTasksInPendingAcquisitionQueue();
                 promise.setSuccess(null);
-
             } else {
-
+                // TODO (DANOBLE) Is this check for an IllegalArgumentException required?
+                //  Explain here, if so; otherwise remove the check.
```
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java:425]: I'd like to leave this in the code for now due to my uncertainty around why the netty devs put this into the base class from which `RntbdClientChannelPool` used to inherit.

[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java:425]: I made a note to remove it with my next PR.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java [hunk 6]
```diff
@@ -27,172 +34,253 @@
...
+//                            Duration.ofNanos(elapsedTimeInNanos),
+//                            Duration.ofNanos(idleEndpointTimeoutInNanos));
+//                    }
+//                    endpoint.close();
+//                    return;
+//                }
+//
+//                this.runTasksInPendingAcquisitionQueue();
+//
+//            }, requestTimerResolutionInNanos, requestTimerResolutionInNanos, TimeUnit.NANOSECONDS);
```
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java:194]: Dead code that's now removed on my local branch. We'll get this in on my next PR which will wire up `DirectConnectionConfig`.

File: sdk/cosmos/azure-cosmos-benchmark/src/main/java/com/azure/cosmos/benchmark/AsyncBenchmark.java [hunk 7]
```diff
@@ -83,11 +88,18 @@
         }
 
         try {
-            cosmosAsyncContainer = cosmosAsyncDatabase.getContainer(this.configuration.getCollectionId()).read().block().getContainer();
+            cosmosAsyncContainer = cosmosAsyncDatabase.getContainer(
+                this.configuration.getCollectionId()
+            ).read().doOnError(error ->
+                logger.error("Database {} creation failed due to ", this.configuration.getDatabaseId(), error)
```
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos-benchmark/src/main/java/com/azure/cosmos/benchmark/AsyncBenchmark.java:104]: This one bothers me and I corrected it on my local branch. I would update the PR if the benchmark tool was shipping code.

[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos-benchmark/src/main/java/com/azure/cosmos/benchmark/AsyncBenchmark.java:104]: And I updated the terminology. Collection is now container.

**Comment by kushagraThapar** [comment]
Some query distinct timeouts in the CI: 
```
Method com.azure.cosmos.rx.DistinctQueryTests.queryDistinctDocuments() didn't finish within the time-out 120000
```

Another flaky test : 

```
Method com.azure.cosmos.implementation.ConsistencyTests2.validateSessionTokenAsync() didn't finish within the time-out 480000

```

**Pull request merged**